golang/cosmos/ante/inbound.go

```diff
@@ -0,0 +1,73 @@
+package ante
+
+import (
+	"github.com/armon/go-metrics"
+
+	"github.com/cosmos/cosmos-sdk/telemetry"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	swingtypes "github.com/Agoric/agoric-sdk/golang/cosmos/x/swingset/types"
+)
+
+// TODO: We don't have a more appropriate error type for this.
+var ErrInboundQueueFull = sdkerrors.ErrMempoolIsFull
+
+type inboundAnte struct {
+	sk SwingsetKeeper
+}
+
+// NewInboundDecorator return an AnteDecorator which honors the allowed size of the inbound queue.
+// The swingset message types will consume one allowed entry each, and will be reflected
+// in the action queue when executed.
+func NewInboundDecorator(sk SwingsetKeeper) sdk.AnteDecorator {
+	return inboundAnte{sk: sk}
+}
+
+// AnteHandle implements sdk.AnteDecorator
+func (ia inboundAnte) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+	msgs := tx.GetMsgs()
+	inboundsAllowed := int32(-1)
+	for _, msg := range msgs {
+		switch msg.(type) {
+		case *swingtypes.MsgDeliverInbound,
+			*swingtypes.MsgInstallBundle,
+			*swingtypes.MsgProvision,
+			*swingtypes.MsgWalletAction,
+			*swingtypes.MsgWalletSpendAction:
+			if inboundsAllowed == -1 {
```

Comment:
ah, ok so this queries the state at most once per txn (not once per msg), and even then only if there were any swingset messages in the txn, nice

Reply:
It even breaks off from iterating through all the Msgs once it discovers that it goes over the limit. The could could be a little more clear if I just summed up all the swingset Msgs in the Tx, rather than computing mid-loop, but it's not too awful as it is.